File: apps/windows/winui/Brand.cs
using System;
using System.Text.Json;
using WinColor = Windows.UI.Color;
using GdiColor = System.Drawing.Color;

namespace DontSpeak;

/// <summary>
/// Brand tints from the Rust core via <see cref="Native.BrandColorsJson"/> (same source as
/// <c>macos/.../Brand.swift</c>), with brand-hex fallbacks. Cached once. Exposed as both
/// <c>Windows.UI.Color</c> (WinUI) and <c>System.Drawing.Color</c> (GDI+ tray/overlay).
/// </summary>
internal static class Brand
{
    public static readonly WinColor SeedPurple;   // speaking / icon seed (#5B4397)
    public static readonly WinColor MicOrange;    // recording (#FF9F0A)
    public static readonly WinColor Warning;      // warming/blocked/download + no-target glow

    public static GdiColor SeedPurpleGdi => Gdi(SeedPurple);
    public static GdiColor MicOrangeGdi => Gdi(MicOrange);

    /// <summary>Logs-tab source palette from <c>ds_log_colors_json</c> (shared across platforms);
    /// first-appearance index picks the color. Brand-hex fallback if engine returns "{}".</summary>
    public static readonly WinColor[] LogSourcePalette;
    private static readonly System.Collections.Generic.Dictionary<string, WinColor> LogLevelColors =
        new(StringComparer.Ordinal);

    /// <summary>ERROR/WARN level color from the shared Rust log-colors source; null for INFO/unknown.</summary>
    public static WinColor? LogLevelColor(string level) =>
        LogLevelColors.TryGetValue(level, out var c) ? c : null;

    static Brand()
    {
        // Fallbacks match Brand.swift — used if the engine returns "{}".
        SeedPurple = FromHex("#5B4397");
        MicOrange = FromHex("#FF9F0A");
        Warning = FromHex("#FF9F0A");
        try
        {
            using var doc = JsonDocument.Parse(Native.BrandColorsJson());
            var root = doc.RootElement;
            SeedPurple = Hex(root, "seed_purple", SeedPurple);
            MicOrange = Hex(root, "mic_orange", MicOrange);
            Warning = Hex(root, "warning", Warning);
        }
        catch { /* engine down / malformed → keep fallbacks */ }

        // Fallbacks mirror Rust defaults so coloring works if the engine returns "{}".
        WinColor[] palette =
        {
            FromHex("#8B7BD8"), FromHex("#3FA7A1"), FromHex("#5B8DEF"), FromHex("#4CAF6E"),
            FromHex("#D97FB0"), FromHex("#CB8A3E"), FromHex("#49B6C2"), FromHex("#B07BD8"),
        };
        LogLevelColors["ERROR"] = FromHex("#E84646");
        LogLevelColors["WARN"] = FromHex("#FF9F0A");
        try
        {
            using var doc = JsonDocument.Parse(Native.LogColorsJson());
            var root = doc.RootElement;
            if (root.TryGetProperty("source_palette", out var pal) && pal.ValueKind == JsonValueKind.Array)
            {
                var list = new System.Collections.Generic.List<WinColor>();
                foreach (var item in pal.EnumerateArray())
                    if (item.ValueKind == JsonValueKind.String && item.GetString() is string s)
                        list.Add(FromHex(s));
                if (list.Count > 0) palette = list.ToArray();
            }
            if (root.TryGetProperty("levels", out var lv) && lv.ValueKind == JsonValueKind.Object)
                foreach (var p in lv.EnumerateObject())
                    if (p.Value.ValueKind == JsonValueKind.String && p.Value.GetString() is string s)
                        LogLevelColors[p.Name] = FromHex(s);
        }
        catch { /* engine down / malformed → keep fallback palette + level colors */ }
        LogSourcePalette = palette;
    }

    private static GdiColor Gdi(WinColor c) => GdiColor.FromArgb(c.A, c.R, c.G, c.B);

    private static WinColor Hex(JsonElement e, string k, WinColor fallback) =>
        e.TryGetProperty(k, out var v) && v.ValueKind == JsonValueKind.String && v.GetString() is string s
            ? FromHex(s) : fallback;

    /// <summary>Parse "#RRGGBB" (opaque); magenta on garbage so bad hex is visible.</summary>
    private static WinColor FromHex(string hex)
    {
        var h = hex.TrimStart('#');
        if (h.Length == 6
            && byte.TryParse(h.AsSpan(0, 2), System.Globalization.NumberStyles.HexNumber, null, out var r)
            && byte.TryParse(h.AsSpan(2, 2), System.Globalization.NumberStyles.HexNumber, null, out var g)
            && byte.TryParse(h.AsSpan(4, 2), System.Globalization.NumberStyles.HexNumber, null, out var b))
            return WinColor.FromArgb(0xFF, r, g, b);
        return WinColor.FromArgb(0xFF, 0xFF, 0x00, 0xFF);
    }
}

File: apps/macos/Sources/DontSpeak/Brand.swift
// Brand colors from ds_brand_colors_json (same ABI as Windows — tints can't drift).
// Hardcoded hexes are FFI-failure fallbacks.

import AppKit
import CDontSpeak
import SwiftUI

enum Brand {
    /// Key → "#RRGGBB", once from ds-core.
    private static let colors: [String: String] = {
        guard let json = ffiString(ds_brand_colors_json), let data = json.data(using: .utf8),
            let map = (try? JSONSerialization.jsonObject(with: data)) as? [String: String]
        else { return [:] }
        return map
    }()

    /// "#RRGGBB" → sRGB; nil if malformed. Single parser for brand + log colors.
    static func nsColor(fromHex hex: String?) -> NSColor? {
        guard let hex, hex.hasPrefix("#"), hex.count == 7,
            let v = Int(hex.dropFirst(), radix: 16)
        else { return nil }
        return NSColor(
            srgbRed: CGFloat((v >> 16) & 0xFF) / 255.0,
            green: CGFloat((v >> 8) & 0xFF) / 255.0,
            blue: CGFloat(v & 0xFF) / 255.0,
            alpha: 1.0
        )
    }

    private static func color(_ key: String, fallback: NSColor) -> NSColor {
        nsColor(fromHex: colors[key]) ?? fallback
    }

    static let seedPurple = color(
        "seed_purple", fallback: NSColor(srgbRed: 0.357, green: 0.263, blue: 0.592, alpha: 1.0))

    static let micOrange = color(
        "mic_orange", fallback: NSColor(srgbRed: 1.0, green: 0.624, blue: 0.039, alpha: 1.0))

    static let warning = color(
        "warning", fallback: NSColor(srgbRed: 1.0, green: 0.624, blue: 0.039, alpha: 1.0))

    // MARK: - Logs colors (ds_log_colors_json — shared with Windows)

    private static let logColors: (levels: [String: String], palette: [String]) = {
        guard let json = ffiString(ds_log_colors_json), let data = json.data(using: .utf8),
            let obj = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        else { return ([:], []) }
        let levels = (obj["levels"] as? [String: String]) ?? [:]
        let palette = (obj["source_palette"] as? [String]) ?? []
        return (levels, palette)
    }()

    /// First-appearance palette index; empty → default text color.
    static let logSourcePalette: [NSColor] = logColors.palette.compactMap { nsColor(fromHex: $0) }

    /// ERROR/WARN from shared map; nil for INFO/unknown.
    static func logLevelColor(_ level: String) -> NSColor? {
        nsColor(fromHex: logColors.levels[level])
    }
}

extension Color {
    static let smWarning = Color(nsColor: Brand.warning)
    /// Update pill accent (brand purple).
    static let smSeedPurple = Color(nsColor: Brand.seedPurple)
}

extension Brand {
    /// One roll from ds_random_pastel_wash_json. Nil if FFI fails.
    static func randomPastelWash() -> Color? {
        guard let json = ffiString(ds_random_pastel_wash_json),
            let data = json.data(using: .utf8),
            let obj = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any],
            let r = obj["r"] as? Int, let g = obj["g"] as? Int, let b = obj["b"] as? Int
        else { return nil }
        let a = (obj["a"] as? Double) ?? (obj["a"] as? NSNumber)?.doubleValue ?? 0.30
        return Color(
            .sRGB,
            red: Double(r) / 255.0,
            green: Double(g) / 255.0,
            blue: Double(b) / 255.0,
            opacity: a
        )
    }
}
